src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,485 @@
+        }
+      case _ => throw new RuntimeException("Invalid Node")
+    }
+  }
+
+  private lazy val EmptyTrieHash = hashFn(encodeRLP(Array.emptyByteArray))
+  lazy val getRootHash: Array[Byte] = rootHash.getOrElse(EmptyTrieHash)
+
+  /**
+    * This function obtains the value asociated with the key passed, if there exists one.
+    *
+    * @param key
+    * @return Option object with value if there exists one.
+    * @throws MPTException if there is any inconsistency in how the trie is build.
+    */
+  def get(key: K): Option[V] = {
+    rootHash flatMap { rootId =>
+      val keyNibbles = HexPrefix.bytesToNibbles(bytes = kSerializer.toBytes(key))
+      val rootNode = getNode(rootId, dataSource)
+      get(rootNode, keyNibbles).map(bytes => vSerializer.fromBytes(bytes))
+    }
+  }
+
+  /**
+    * This function inserts a (key-value) pair into the trie. If the key is already asociated with another value it is updated.
+    *
+    * @param key
+    * @param value
+    * @return New trie with the (key-value) pair inserted.
+    * @throws MPTException if there is any inconsistency in how the trie is build.
+    */
+  def put(key: K, value: V): MerklePatriciaTrie[K, V] = {
+    val keyNibbles = HexPrefix.bytesToNibbles(kSerializer.toBytes(key))
+    rootHash map { rootId =>
+      val root = getNode(rootId, dataSource)
+      val NodeInsertResult(newRoot, nodesToRemoveFromStorage, nodesToUpdateInStorage) = put(root, keyNibbles, vSerializer.toBytes(value))
+      val newRootHash = newRoot.hash
+      val newSource = updateNodesInStorage(
+        previousRootHash = getRootHash,
+        newRootHash = newRootHash,
+        newRoot = Some(newRoot),
+        toRemove = nodesToRemoveFromStorage,
+        toUpdate = nodesToUpdateInStorage,
+        dataSource = dataSource,
+        hashFn = hashFn)
+      new MerklePatriciaTrie(Some(newRootHash), newSource, hashFn)(kSerializer, vSerializer)
+    } getOrElse {
+      val newRoot = LeafNode(keyNibbles, vSerializer.toBytes(value), hashFn)
+      val newRootHash = newRoot.hash
+      new MerklePatriciaTrie(Some(newRootHash),
+        updateNodesInStorage(getRootHash, newRootHash, Some(newRoot), Seq(), Seq(newRoot), dataSource, hashFn),
+        hashFn)
+    }
+  }
+
+  /**
+    * This function deletes a (key-value) pair from the trie. If no (key-value) pair exists with the passed trie then there's no effect on it.
+    *
+    * @param key
+    * @return New trie with the (key-value) pair associated with the key passed deleted from the trie.
+    * @throws MPTException if there is any inconsistency in how the trie is build.
+    */
+  def remove(key: K): MerklePatriciaTrie[K, V] = {
+    rootHash map { rootId =>
+      val keyNibbles = HexPrefix.bytesToNibbles(bytes = kSerializer.toBytes(key))
+      val root = getNode(rootId, dataSource)
+      remove(root, keyNibbles) match {
+        case NodeRemoveResult(true, Some(newRoot), nodesToRemoveFromStorage, nodesToUpdateInStorage) =>
```

Comment:
about second parameter it is called here `newRoot` but in case class this field is named `maybeNewChild`
https://github.com/input-output-hk/etc-client/pull/10/files#diff-2c1aab8e8ca156fbe430bbb586ecfc01R17
should it be that way?